idaes/dmf/tables.py

```diff
@@ -0,0 +1,148 @@
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+###############################################################################
+"""
+Table handling for DMF
+"""
+# stdlib
+from typing import List, Tuple, Dict
+import re
+
+# ext
+import pandas as pd
+
+__author__ = "Dan Gunter"
+
+
+class Table:
+    """Represent a table stored in the DMF.
+    """
+    def __init__(self):
+        self._data = pd.DataFrame({})
+        self._units = {}
+        self._filepath = None
+
+    @property
+    def data(self) -> pd.DataFrame:
+        return self._data
+
+    @property
+    def units(self) -> Dict[str, str]:
+        return self._units.copy()
+
+    def read_csv(self, filepath, **kwargs) -> None:
+        """Read the table from a CSV file using pandas' `read_csv()`.
+        See `Pandas read_csv docs
+        <https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html>`_
+        for details.
+
+        Existing table will be replaced.
+
+        Args:
+            filepath: Any valid first argument to pandas `read_csv`
+            kwwargs: Keyword arguments passed to pandas `read_csv`
+
+        Returns:
+            None
+        """
+        self._data = pd.read_csv(filepath, **kwargs)
+        self._extract_units()
+        self._filepath = filepath
+
+    def read_excel(self, filepath, **kwargs) -> None:
+        """Read the table from a CSV file using pandas' `read_excel()`.
+        See `Pandas read_excel docs
+        <https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html>`_
+        for details.
+
+        Existing table will be replaced.
+
+        Args:
+            filepath: Any valid first argument to pandas `read_excel`
+            **kwargs: Keyword arguments passed to pandas `read_excel`
+
+        Returns:
+            None
+
+        Raises:
+            ValueError, if more than one Excel sheet is returned
+        """
+        data = pd.read_excel(filepath, **kwargs)
+        if isinstance(data, dict):
+            raise ValueError(f"Read from excel file must return a single sheet, "
+                             f"but sheet_name='{kwargs.get('sheet_name', '?')}' "
+                             f"returned {len(data)} sheets: {list(data.keys())}")
+        self._data = data
+        self._extract_units()
+        self._filepath = filepath
+
+    def _extract_units(self):
+        new_names, units_dict = {}, {}
+
+        for name in self._data.columns:
+            base_name, units = self._split_units(name)
+            new_names[name] = base_name
+            units_dict[base_name] = units
+
+        self._data.rename(columns=new_names, inplace=True)
+        self._units = units_dict
+
+    @staticmethod
+    def _split_units(name) -> Tuple[str, str]:
```

Comment:
Like most people, I struggle to understand regular expressions. Could you comment on what form this expects the units to be in?

For reference, in the set of tables of interest to me at the moment some example headers are (in my current formatting):

T [C]
b [mol/kg]
H0/RT [-]
10^3 A_k [cc.kg^1/2.mol^2/3/bar]

The bit between square brackets are the units, the rest is the variable name. Note also that in some cases we have a multiplier on a variable as part of the name (e.g. 10^3 A_k), which I haven't thought through much yet.

Reply:
Right, square brackets. Or missing entirely is also OK (same as `[-]`) as long as you don't use a square bracket in the units themselves, you can put pretty much anything else in there